### ISSUE TITLE
Tracking Issue for Vec::push_within_capacity

### ISSUE THREAD

**Issue by the8472** [rust-lang/rust#100486]
Feature gate: `#![feature(vec_push_within_capacity)]`

This is a tracking issue for `Vec::push_within_capacity`

It enables pushing into a Vec if there is any capacity left and otherwise returns the value. I.e. it avoids implicit resizing. This can be useful when one wants to explicitly control it (e.g. via `try_reserve`) or uses a Vec in an unsafe manner where moving the elements would be a problem.

### Public API

```rust
// alloc::Vec

impl Vec<T, A> {
    pub fn push_within_capacity(&mut self, value: T) -> Result<(), T>;
}
```

### Steps / History

- [ ] Implementation: #89123
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- [ ] naming (e.g. `push_in_capacity`)
- [ ] may become redundant due to larger-scale work in https://github.com/rust-lang/rfcs/pull/3271

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by em-tg** [comment]
Would it be possible for `push_within_capacity` to take `self` by shared reference?  If the backing store is never re-allocated then calling this wouldn't invalidate any references to the `Vec`'s contents, I'd think.

**Comment by the8472** [comment]
It still needs to bump the length field.

**Comment by em-tg** [comment]
Wouldn't wrapping the length in `UnsafeCell` work for that?  The public api only exposes the `len()` getter.

**Comment by em-tg** [comment]
Ah wait I think I see the issue.  You'd lose `Sync` by doing that.  Sorry, still pretty new to Rust and just thought I'd ask!

**Comment by tiby312** [comment]
I think this is a great feature. I often find myself just assuming my buffer reserving code is working since it is a hard bug to catch when it's not working given push automatically reserves.

**Comment by krtab** [comment]
Hi!

I think this could enter its FCP:
- https://github.com/rust-lang/rfcs/pull/3271 is not merged. Parts of it were released as a crate, part of it has been delegated to keyword genericity (so, pretty far away)
- The code of #89123 has been in nightly for a year and received no criticisms afaict
- This API can prove very useful to add optimization hints by marking its `Err` return value as unreachable

**Comment by kornelski** [comment]
I find it very useful, and have similar helpers in my codebases. It generates much leaner code than the usual `push` when the optimizer can't be convinced that the capacity is sufficient (e.g. when it's calculated in a non-trivial way, or the vec-appending code isn't all inlined). It has a useful side effect of verifying that a sufficient capacity has been precalculated,  and there aren't unexpected reallocations.

However, just `push` is not enough. I'm missing similar functionality for `extend_from_slice` — `_within_capacity`.

**Comment by elichai** [comment]
I hope this isn't completely useful,
But is there some kind of mechanism we can use to allow all these `try_` functionality without adding dozens of `try_` in all collections, could being generic over the effect of `try` allow something like this?

**Comment by Darksonn** [comment]
How about `insert_within_capacity`? It's very non-trivial to reimplement `insert` without an allocation codepath.

**Comment by the8472** [comment]
Sounds interesting, though a bit niche (just like this method). If you have a use-case you should file an ACP for it.
Ditto @kornelski 

If you several a bunch of methods like this in mind then we could perhaps bundle them to see if there are some overarching things to consider.

**Comment by kornelski** [comment]
There are two ways to implement such functionality, and I'm not sure which way to propose:

1. Add more methods on `Vec` with a `_within_capacity` version. I think this would work well, but I've got a lot of pushback about adding `try_` versions of methods. Rust doesn't have a feature like Swift's function naming/overloading to keep all of these functionality tweaks elegantly together.

2. A reference type like `FixedCapacityVecView<'_>` that imitates interface of `Vec`, but without reallocation. This seems like a big addition, especially if it will duplicate most of Vec's interface, but most users will probably only need `push` and `extend`.

**Comment by GoldsteinE** [comment]
This `FixedCapacityVecView<'_>` is kinda similar to `std::io::BorrowedBuf<'_>` in both its API and idea.

**Comment by kornelski** [comment]
Although it's somewhat similar, `BorrowedBuf` is not a replacement:

* it only supports `u8`, no other types.
* it tracks `initialized` and `filled` levels separately, which is unnecessary for `Vec`. Just an extra field may seem like a small issue, but `Vec` had issues with missed optimizations and bloated code when LLVM lost track of `len <= capacity` relationship, and I'm afraid that tracking relationships betwee `len`, `capacity`, `spare_capacity_slice.len()`, `filled`, and `initialized` all at once may be hard for LLVM in hot loops. I use push in capacity mainly to reduce branches and code size.
* it doesn't seem to have a way to integrate with a `Vec` in safe code. User needs to use `vec.set_len()` to update it with the `filled` len. This actually seems like a [flaw in `BorrowedVec`](https://github.com/rust-lang/rust/issues/117693#issuecomment-2150398589) to me.

**Comment by the8472** [comment]
I assume that was only intended to point at some similar-not-identical prior art, not suggesting that BorrowedBuf can be reused for the purpose.

> Add more methods on Vec with a _within_capacity version. I think this would work well, but I've got a lot of pushback about adding try_ versions of methods. 

Yeah, imo this is the biggest obstacle to getting anything done. It's either a lot new awkwardly named methods, view types or some generics/associated type shenanigans.
And some of the APIs require new design because they're not just 1:1 changes. Note how `push_within_capacity` has a return type, unlike push.

Thinking about design I now realize that there's an alternative, it could take a `&mut MaybeUninit<T>` and not move the value if the capacity is full. Would that bring any perf benefits, maybe for large values?

And that's why it's often suggested to explore these thing as a crate outside of std first.

**Comment by mcroomp** [comment]
How about something like this:
```
pub fn needs_to_grow<T>(&self, additional: usize) -> bool {
    additional > self.capacity().wrapping_sub(self.len())
}

```

it makes it easy to solve both problems... the trick is that the implementation must match exactly to what RawVec does, otherwise the optimizer doesn't realize that the operations are identical.

**Comment by kornelski** [comment]
I don't think it's important to return elements that couldn't fit in the capacity. This isn't like I/O functions or concurrent queues where you can't predict whether insert will succeed or not. When you want to keep the elements that can't be inserted, it's sufficient to check the capacity before push/extend.

**Comment by mcroomp** [comment]
> I don't think it's important to return elements that couldn't fit in the capacity. This isn't like I/O functions or concurrent queues where you can't predict whether insert will succeed or not. When you want to keep the elements that can't be inserted, it's sufficient to check the capacity before push/extend.

The main problem is that obvious `x.len() < x.capacity()` will not convince the optimizer not to check twice, you have to do `additional > self.capacity().wrapping_sub(self.len())` to match the check that RawVec makes.

**Comment by hanna-kruppe** [comment]
Returning the value that couldn’t be pushed/inserted/etc. only seems to matter if:

1. The element type isn’t Copy, *and*
2. The code using `*_within_capacity` can do something useful with the excess value after failing to put it in the Vec, *and*
3. It’s important that the capacity is only checked once for runtime efficiency or binary size.

I struggle to imagine a use case with these properties that isn’t isomorphic to “generic `try_{push,insert,…}` that handles OOM by returning Err”. The other uses I know of only involve Copy types (e.g., T = u8 for byte centric I/O without `std::io`), or have sufficient capacity as logical invariant (i.e., if the within_capacity call fails, that’s not meaningfully recoverable).

I guess if you use a Vec as a fixed size buffer of non-Copy elements and want to flush the buffer to another stage of processing whenever it gets full? This needs `push_within_capacity` (or eat the potential cost of a manual capacity check) but not any other methods (that imply a capacity check). A moving `extend_from_slice` analogue is sometimes useful, but if you want to do better than pushing one element at a time then the technical challenge isn’t “de-duplicate a capacity check” but rather arranging a right-sized `memcpy` from the source into your buffer (repeatedly until the source is exhausted). This can be done today with `spare_capacity_mut` without having to assume anything undocumented about Vec.

If any use case for more than `push_within_capacity` amounts to tackling the “zoo of `Vec::try_*`” challenge, then I’d rather see just `push_within_capacity` stabilized on its own merits.

**Comment by LegionMammal978** [comment]
Returning the value can be useful in cases where immediately dropping it would be unsound. E.g., I could write a `VecCell<T>` type that logically acts like a `Cell<Vec<T>>`, offering safe operations on its items ([Rust Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=50a97cb502a3a1d57a886287df1ecd30)).

Even a panicking `VecCell::push()` method cannot be built directly on `Vec::push()`, since on failure, `Vec::push()` could (a) drop a value with a destructor reentering a `VecCell` method, or (b) call a panic hook reentering a `VecCell` method, either of which would be unsound. Instead, `VecCell::push()` must call either `Vec::push_within_capacity()` + `Vec::try_reserve()` + `panic!()`, or a hypothetical `Vec::try_push()` + `panic!()`, so that the drop and panic only occur once the dangerous `&mut Vec<T>` is no longer in use.

In other words, contexts with restricted drops or panics do need a returned value to extricate from that context, even if they are not outwardly "isomorphic to 'generic `try_{push,insert,…}` that handles OOM by returning Err'."

**Comment by hanna-kruppe** [comment]
In brief (because this is getting quite far afield from `push_within_capacity`): yes, you have to be hyper-vigilant when you're writing code that can't tolerate unwinding. That's why it's important to minimize the amount of code transitively used in such regions, including by relying on the absolute minimum amount of external code (including std) that may not be carefully written with the same concerns in mind. Your expectation that all std APIs guarantee whether they may allocate/panic/etc. does not match current reality and if you want to try to change that, this isn't the venue for that.
